Ability to use NeoPixel leds

hallard opened the pull request:
All by boards uses neopixel, this one allow to use them, may be implemented into yasolr :-)

- Added NeoPixel possible use (even both same time)
- Added NeoPixel Example
- Added Adafruit NeoPixel Library

For now straight compatibility just drive one color on each led Green, Yellow and Red, if needed I'll add ability to traffic light only with one Led.

mathieucarbou commented:
Thanks! I wasn't expecting some PR on such a small project :P I will review!

I have updated the CI files for now (which conflicts a little)

mathieucarbou commented on examples/TrafficLightNeoPixel/TrafficLightNeoPixel.ino:
```diff
@@ -0,0 +1,56 @@
+//   NEO_GRB     Pixels are wired for GRB bitstream (most NeoPixel products)
+//   NEO_RGB     Pixels are wired for RGB bitstream (v1 FLORA pixels, not v2)
+//   NEO_RGBW    Pixels are wired for RGBW bitstream (NeoPixel RGBW products)
+constexpr neoPixelType NEOPIXEL_TYPE = NEO_GRB + NEO_KHZ800 ;
```
I would call if `NEOPIXEL_FLAGS`

mathieucarbou commented on examples/TrafficLightNeoPixel/TrafficLightNeoPixel.ino:
```diff
@@ -0,0 +1,56 @@
+  lights.begin( NEOPIXEL_PIN, NEOPIXEL_COUNT, NEOPIXEL_TYPE ); 
+
+  // Set BRIGHTNESS to about 1/5 (max = 255)
+  lights.setBrightness(50); 
```
here I would prefer to have a float `[0.0, 1.0]`, because it will cover any use case (even future ones).
the limit of 255 is implementation specific for now (even if we could argue that it matches a 8 bit brightness) so the level could then be found by multiplying the % with the limit.

hallard commented:
> Thanks! I wasn't expecting some PR on such a small project :P I will review!
> 
> I have updated the CI files for now (which conflicts a little)

hidden goal is to be able to use them with small boards for yasolr like this one all small connectors to connect relays :-)


![trim 80E56AB9-6091-43F1-8E78-3FCAFE09B28A](https://github.com/user-attachments/assets/7d3a1322-01be-43da-a377-28dc9b08078f)

mathieucarbou commented on platformio.ini:
```diff
@@ -6,8 +6,10 @@ build_flags =
+;  -D MYCILA_NEOPIXEL_SUPPORT
 lib_deps = 
   bblanchon/ArduinoJson @ 7.2.0
+  adafruit/Adafruit NeoPixel @ 1.12.3
```
I've reworked the CI and pio file in order to be able to add an isolated section for neopixel (like it is done with json) in order to test it better in CI

hallard replied:
Not sure if I need to do something on this one?

mathieucarbou replied:
I updated the CI part ;-)

mathieucarbou commented on src/MycilaTrafficLight.cpp:
```diff
@@ -54,6 +54,29 @@ void Mycila::TrafficLight::begin(const int8_t greenPin, const int8_t yellowPin,
 }
 
+#ifdef MYCILA_NEOPIXEL_SUPPORT
+void Mycila::TrafficLight::begin(const int16_t neoPin, const uint16_t neoCount, const neoPixelType neoType) {
```
`int8_t neoPin` would be enough for a pin ?

hallard replied:
Sure, as behind the scene I'm calling Adafruit lib, I just let same type they made

mathieucarbou commented on src/MycilaTrafficLight.cpp:
```diff
@@ -54,6 +54,29 @@ void Mycila::TrafficLight::begin(const int8_t greenPin, const int8_t yellowPin,
+#ifdef MYCILA_NEOPIXEL_SUPPORT
+void Mycila::TrafficLight::begin(const int16_t neoPin, const uint16_t neoCount, const neoPixelType neoType) {
+  // In case we already have a NeoPixel object, delete it
+  if (_strip != nullptr) {
+    delete _strip;
+    _strip = nullptr;
+  }
```
this lib is the only one which didn't have a `_enabled` flag to check its lifecycle. I think it should have one now. I will add it in main in the same way I use to have it in all my libs so that it stays consistant.

hallard commented:
> I have updated the CI files for now (which conflicts a little)

Sorry I'm far from expert in PR/Rebase/Conflict, don't know what I need to do on this one so solve conflict, I think I'll let you hand on this.

mathieucarbou reviewed (commented):
I have a few minor comments.

I have pushed in main a refresh for CI and lifecycle, to align this lib with all the other ones I maintain.

- you will be able to rebase
- use the `_gpioEnabled` flag to check for lifecycle (end)
- also review the cpplint checks 

I will merge after.

Do you know some NeoPixel leds with only 3 leds ? I saw strips and single led, but did not fidn a set of 3 ?

Thanks 👍

mathieucarbou commented:
> > I have updated the CI files for now (which conflicts a little)
> 
> Sorry I'm far from expert in PR/Rebase/Conflict, don't know what I need to do on this one so solve conflict, I think I'll let you hand on this.

ok I will grab your changes then :-)

mathieucarbou commented:
@hallard : please let me know when I can rebase your PR branch on main and add the CI :-) 
I will need to force push into your branch for that (I see I should be able to do it).

hallard commented:
@mathieucarbou should be good for now :-)

hallard commented:
> * use the `_gpioEnabled` flag to check for lifecycle (end)

Not sure how to handle this one, in fact I though you could have both call to `begin()` one for gpio other for neopixel (all leds would be synched of course in this case)
Imagine call 1st for neopixel then call second with wrong ping, then 2nd set `_gpioEnabled` to fasle and by the way disable 1st one. I don't think it's the expected result.

May be we could handle another flag as `_neopixelEnabled` or use gpio mode OR neopixel but not both same time.

> Do you know some NeoPixel leds with only 3 leds ? I saw strips and single led, but did not fidn a set of 3 ?

Nope usually I cut strip or use this kind of awesome, lot of nice models 

![image](https://github.com/user-attachments/assets/bce84a8e-7885-4e6c-9fc9-ce7eb8133c1d)

got theese one in 5V can't find them now, looks like all 12V
![image](https://github.com/user-attachments/assets/7c5eb692-76b2-45ff-8d6f-97add71f7a08)


https://www.aliexpress.com/w/wholesale-20pcs-rgb-pixel-led-ws2811-5V.html

mathieucarbou commented:
Thanks for the hints! I will have a look. These are funny: I am so busy with other things... I should pause for a few weeks at some point just to make something fun for the kids with that ;-)

mathieucarbou commented:
@hallard : I have rebased the PR, updated the CI and did some updates on the code for a better integration with what was existing. Would you be able to test ? If it works, I will merge and release.

Thanks for this addition!

hallard commented:
> @hallard : I have rebased the PR, updated the CI and did some updates on the code for a better integration with what was existing. Would you be able to test ? If it works, I will merge and release.
> 
> Thanks for this addition!

@mathieucarbou I'm very sorry, I'm totally lost, between my fork, rebase from main of my branch of for upstream, and merge sometime git drives me mad just for few line of code. I propose you to merge, then I will delete my branch and sync my fork and test. If any issue I will create a new branch if ok to you 

hallard commented:
> Thanks for the hints! I will have a look. These are funny: I am so busy with other things... I should pause for a few weeks at some point just to make something fun for the kids with that ;-)

I know what it is ... Can send you some if needed, got some of old projects, looks great to put in Solar Router Enclosure 

![image](https://github.com/user-attachments/assets/f0e12c73-1fde-44f1-8d5d-40e2520d8865)

hallard commented:
@mathieucarbou looks like I succeeded to resolv conflicts on my side, but since I can't build anymore nor test.
So for now I'm giving up, I don't need to spend more time on trying mess on my side than time to code feature.
Don't blame you of course, It's not the first time, each time I'm forking and creating branch while upstream is moving it's a mess on my side with rebase, merge, resync, conflict. I never have any issue on my repo or repos with permission, always on forked ones but may be I'm not smart enough for that.

mathieucarbou commented:
> > @hallard : I have rebased the PR, updated the CI and did some updates on the code for a better integration with what was existing. Would you be able to test ? If it works, I will merge and release.
> > Thanks for this addition!
> 
> @mathieucarbou I'm very sorry, I'm totally lost, between my fork, rebase from main of my branch or for upstream, and merge. Sometimes git drives me mad just for few lines of code. I propose you to merge, then I will delete my branch and sync my fork and test. If any issue, I will create a new branch to fix if ok to you?

No worry I have the local copy. 
You could have done a got fetch and git reset --hard to set you local copy up to date

mathieucarbou commented:
v1.1.0 is released with he PR: I will integrate it in YaSolR now.

hallard commented:
Just tested, it works.

Just you let know that default compile program in `platformio.ini` is set to `TrafficLightNeoPixel` not sure it's intentional or not.

mathieucarbou commented:
Thanks!
Ye sit does not matter: this is for the devs ;-) the ci will force and execute all examples.

mathieucarbou commented:
Oh! I added a Todo also in the kanban board of YaSolR (project tab) to add support for that once I will have the time.
This is low priority for now though.
